activemodel/lib/active_model/attribute_methods.rb

```diff
@@ -322,157 +301,32 @@ def define_attribute_method(attr_name)
-        # this cache).
-        def attribute_method_matchers_cache
-          @attribute_method_matchers_cache ||= Concurrent::Map.new(initial_capacity: 4)
-        end
-
-        def attribute_method_matchers_matching(method_name)
-          attribute_method_matchers_cache.compute_if_absent(method_name) do
-            # Must try to match prefixes/suffixes first, or else the matcher with no prefix/suffix
-            # will match every time.
-            matchers = attribute_method_matchers.partition(&:plain?).reverse.flatten(1)
-            matchers.map { |method| method.match(method_name) }.compact
-          end
-        end
-
-        # Define a method `name` in `mod` that dispatches to `send`
-        # using the given `extra` args. This falls back on `define_method`
-        # and `send` if the given names cannot be compiled.
-        def define_proxy_call(include_private, mod, name, send, *extra)
-          defn = if NAME_COMPILABLE_REGEXP.match?(name)
-            "def #{name}(*args)"
-          else
-            "define_method(:'#{name}') do |*args|"
-          end
-
-          extra = (extra.map!(&:inspect) << "*args").join(", ".freeze)
-
-          target = if CALL_COMPILABLE_REGEXP.match?(send)
-            "#{"self." unless include_private}#{send}(#{extra})"
-          else
-            "send(:'#{send}', #{extra})"
-          end
+      def attribute_methods_builder_class
+        AttributeMethodsBuilder
+      end
 
-          mod.module_eval <<-RUBY, __FILE__, __LINE__ + 1
-            #{defn}
-              #{target}
+      private
+        def _attribute_methods_builder
+          @attribute_methods_builder ||=
+            attribute_methods_builder_class.new.tap do |builder|
+              self.attribute_methods_builders += [builder]
+              builder.apply(self)
             end
-          RUBY
         end
 
-        class AttributeMethodMatcher #:nodoc:
-          attr_reader :prefix, :suffix, :method_missing_target
-
-          AttributeMethodMatch = Struct.new(:target, :attr_name, :method_name)
-
-          def initialize(options = {})
-            @prefix, @suffix = options.fetch(:prefix, ""), options.fetch(:suffix, "")
-            @regex = /^(?:#{Regexp.escape(@prefix)})(.*)(?:#{Regexp.escape(@suffix)})$/
-            @method_missing_target = "#{@prefix}attribute#{@suffix}"
-            @method_name = "#{prefix}%s#{suffix}"
-          end
-
-          def match(method_name)
-            if @regex =~ method_name
-              AttributeMethodMatch.new(method_missing_target, $1, method_name)
-            end
-          end
-
-          def method_name(attr_name)
-            @method_name % attr_name
-          end
-
-          def plain?
-            prefix.empty? && suffix.empty?
-          end
+        # Note: method_missing overrides are only defined once the builder is
+        # first used, to avoid unnecessarily increasing execution time of
+        # +respond_to?+ in case the builder has no matchers on it.
```

Comment:
Could this be left up to the builder itself, to define method_missing when it actually has a matcher?

Reply:
Yes it could, and I hesitated a lot about this. I opted to let AM::AttributeMethods decide when to define them so that the builders could in theory be used without the method missing overrides, but maybe that's overthinking it.

Reply:
Actually the reason I don't do this is because there is a situation where we want to include a builder with no method missing/respond_to on it. This is in AR::Base, when defining prefixes and suffixes. We duplicate this builder in inherited classes (i.e. all model classes) and then define mm/respond_to on this duplicated module, but not on the original one since this would be unnecessary and cause an unnecessary code path.

This is actually the most tricky part of this code, and it's tricky mainly because I want to: 1) comply with the current API, 2) decouple core logic into the builders, and 3) avoid any extra method calls. Satisfying these three together is hard, but can be done at the cost of these kinds of subtle considerations.

Reply:
Does the module need to be included at all in those cases? It needs to be there to gather definitions, and to be duplicated and included by non-abstract children... but maybe doesn't need to be in the inheritance chain?

Reply:
I've lost track of which line you're referencing, but for ActiveRecord, it needs to be included in every non-abstract, non-base class in order to define methods (attribute methods + method_missing/respond_to) on that class. For abstract and base classes, we don't need to include it since we never instantiate those classes (so we don't actually need the methods).

For Active*Model*, nothing is included until you actually call a prefix/suffix/define method. This by the way implies that one subtle corner case of the current code is not maintained: the case where you `include ActiveModel::AttributeMethods`, define an `attributes` method with some attributes, and then try calling one of the keys as a method. Currently this will be caught by `method_missing`, but with this PR change, mm/respond_to will not be defined *until* you actually add at least one prefix/suffix/affix *or* call `define_attribute_methods`.

This was a trade-off because I consider this a very rare case and it significantly complicates implementation, for little benefit (there is no impact on AR::Base models from this).